File: backend/src/main/java/com/smartcampus/backend/service/EmailService.java
package com.smartcampus.backend.service;

import jakarta.mail.MessagingException;
import jakarta.mail.internet.MimeMessage;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.mail.javamail.JavaMailSender;
import org.springframework.mail.javamail.MimeMessageHelper;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;

/**
 * Handles all outbound email for the platform.
 * Methods are @Async — email sending never blocks the HTTP request thread.
 */
@Service
public class EmailService {

    private static final Logger log = LoggerFactory.getLogger(EmailService.class);

    private final JavaMailSender mailSender;

    public EmailService(JavaMailSender mailSender) {
        this.mailSender = mailSender;
    }

    /**
     * Sends the 6-digit OTP verification email to a newly registered user.
     * Called asynchronously so registration response returns instantly.
     *
     * @param toEmail  Recipient email address
     * @param fullName User's full name (for personalisation)
     * @param otpCode  6-digit OTP string
     */
    @Async
    public void sendOtpEmail(String toEmail, String fullName, String otpCode) {
        try {
            MimeMessage message = mailSender.createMimeMessage();
            MimeMessageHelper helper = new MimeMessageHelper(message, true, "UTF-8");

            helper.setTo(toEmail);
            helper.setSubject("Your SmartCampus Verification Code: " + otpCode);
            helper.setText(buildOtpEmailHtml(fullName, otpCode), true); // true = HTML

            mailSender.send(message);
            log.info("OTP email sent to {}", toEmail);

        } catch (MessagingException e) {
            // Log but don't throw — a failed email shouldn't crash the registration
            // response.
            // The user can use the resend endpoint.
            log.error("Failed to send OTP email to {}: {}", toEmail, e.getMessage());
        }
    }

    /**
     * Builds a professional HTML email body for the OTP.
     * Inline styles used because most email clients strip <style> tags.
     */
    private String buildOtpEmailHtml(String fullName, String otpCode) {
        // Split OTP into individual digit spans for the visual block style
        StringBuilder digitSpans = new StringBuilder();
        for (char c : otpCode.toCharArray()) {
            digitSpans.append(
                    "<span style=\"display:inline-block; width:48px; height:56px; " +
                            "line-height:56px; text-align:center; font-size:28px; font-weight:700; " +
                            "color:#1a1a2e; background:#f0f4ff; border:2px solid #d0d9ff; " +
                            "border-radius:8px; margin:0 4px;\">" + c + "</span>");
        }

        return """
                <!DOCTYPE html>
                <html>
                <body style="margin:0; padding:0; background-color:#f5f7fa; font-family: 'Segoe UI', Arial, sans-serif;">
                  <table width="100%%" cellpadding="0" cellspacing="0" style="background:#f5f7fa; padding: 40px 0;">
                    <tr>
                      <td align="center">
                        <table width="520" cellpadding="0" cellspacing="0"
                               style="background:#ffffff; border-radius:16px; overflow:hidden;
                                      box-shadow: 0 4px 24px rgba(0,0,0,0.08);">

                          <!-- Header -->
                          <tr>
                            <td style="background: linear-gradient(135deg, #667eea 0%%, #764ba2 100%%);
                                       padding: 32px 40px; text-align:center;">
                              <h1 style="margin:0; color:#ffffff; font-size:24px; font-weight:700;
                                         letter-spacing:-0.5px;">🍔 SmartCampus</h1>
                              <p style="margin:6px 0 0; color:rgba(255,255,255,0.85); font-size:14px;">
                                Campus Food Ordering Platform
                              </p>
                            </td>
                          </tr>

                          <!-- Body -->
                          <tr>
                            <td style="padding: 40px 40px 32px;">
                              <p style="margin:0 0 8px; font-size:18px; font-weight:600; color:#1a1a2e;">
                                Hi %s! 👋
                              </p>
                              <p style="margin:0 0 28px; font-size:15px; color:#555; line-height:1.6;">
                                Welcome to SmartCampus! Please verify your email address
                                to activate your account and start ordering.
                              </p>

                              <!-- OTP block -->
                              <div style="text-align:center; margin: 0 0 28px;">
                                <p style="margin:0 0 16px; font-size:13px; color:#888;
                                          text-transform:uppercase; letter-spacing:1px; font-weight:600;">
                                  Your verification code
                                </p>
                                <div>%s</div>
                              </div>

                              <!-- Expiry note -->
                              <div style="background:#fff8e1; border-left:4px solid #ffc107;
                                          border-radius:4px; padding: 12px 16px; margin-bottom:28px;">
                                <p style="margin:0; font-size:13px; color:#856404;">
                                  ⏱ This code expires in <strong>10 minutes</strong>.
                                  Do not share it with anyone.
                                </p>
                              </div>

                              <p style="margin:0; font-size:14px; color:#888; line-height:1.6;">
                                If you didn't create a SmartCampus account, you can safely ignore this email.
                              </p>
                            </td>
                          </tr>

                          <!-- Footer -->
                          <tr>
                            <td style="background:#f8f9fa; padding: 20px 40px; text-align:center;
                                       border-top: 1px solid #eee;">
                              <p style="margin:0; font-size:12px; color:#aaa;">
                                © SmartCampus Platform &nbsp;·&nbsp; This is an automated message, please do not reply.
                              </p>
                            </td>
                          </tr>

                        </table>
                      </td>
                    </tr>
                  </table>
                </body>
                </html>
                """
                .formatted(fullName, digitSpans.toString());
    }
}

File: backend-fastapi/app/services/email_service.py
"""
Email service — async SMTP send for email-verification OTPs.

Source of truth: EmailService.java (Spring Boot).

This service is EXCLUSIVELY for the EMAIL-VERIFICATION OTP system (registration
flow, DB-persisted, time-expiring tokens — see otp_service.py + EmailOtpToken).
It must NEVER be shared with the pickup OTP system, which is HMAC-derived and
never expires (spec §4.2).

FIRE-AND-FORGET SEMANTICS
-------------------------
The Java original annotated sendOtpEmail with @Async and swallowed
MessagingException (log, don't throw). A failed email therefore never crashes
the registration response; the user can retry via the resend endpoint.

We reproduce that here: otp_service.generate_and_send_otp awaits
send_otp_email directly (not via FastAPI BackgroundTasks), and every exception
inside this function is caught and logged — never re-raised. Direct await is
chosen over BackgroundTasks because Java's @Async dispatched the email BEFORE
the outer register transaction's throw — BackgroundTasks do not fire on
exception, which would drop the email in the duplicate-pending 409 path.

DEV / CI MODE
-------------
If MAIL_USERNAME is empty (no SMTP credentials configured) the function logs a
warning and returns immediately, matching the Java silent-failure behaviour.
This lets registration work end-to-end in dev/CI without a real Gmail account;
the OTP is still visible in the server logs via the DEBUG line in otp_service.
"""
from __future__ import annotations

import logging
from email.message import EmailMessage

import aiosmtplib

from app.config import settings

logger = logging.getLogger(__name__)


def _build_otp_email_html(full_name: str, otp_code: str) -> str:
    """
    Port of EmailService.buildOtpEmailHtml — identical HTML structure & inline
    styles so the rendered email looks exactly like the Spring Boot version.

    Layout: gradient header "SmartCampus", greeting with the user's full name,
    six digit-span boxes rendering the OTP, a 10-minute expiry note, an
    "if you didn't request this, ignore this email" line, and a footer.
    """
    digit_spans = "".join(
        '<span style="display:inline-block; width:48px; height:56px; '
        "line-height:56px; text-align:center; font-size:28px; font-weight:700; "
        "color:#1a1a2e; background:#f0f4ff; border:2px solid #d0d9ff; "
        'border-radius:8px; margin:0 4px;">' + c + "</span>"
        for c in otp_code
    )

    return """<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>SmartCampus Verification Code</title>
</head>
<body style="margin:0; padding:0; background:#f4f4f7; font-family:'Segoe UI', Arial, sans-serif;">
<table role="presentation" width="100%%" cellpadding="0" cellspacing="0" style="background:#f4f4f7; padding:32px 0;">
<tr>
<td align="center">
<table role="presentation" width="480" cellpadding="0" cellspacing="0" style="background:#ffffff; border-radius:12px; overflow:hidden; box-shadow:0 4px 16px rgba(0,0,0,0.08);">
<tr>
<td style="background:linear-gradient(135deg,#667eea 0%%,#764ba2 100%%); padding:32px 40px; text-align:center;">
<h1 style="margin:0; color:#ffffff; font-size:24px; font-weight:700; letter-spacing:0.5px;">SmartCampus</h1>
</td>
</tr>
<tr>
<td style="padding:40px 40px 24px 40px;">
<p style="margin:0 0 8px 0; font-size:18px; font-weight:600; color:#1a1a2e;">Hi %s! Welcome to SmartCampus.</p>
<p style="margin:0 0 24px 0; font-size:15px; color:#444; line-height:1.6;">
Your verification code is:
</p>
<div style="text-align:center; margin:24px 0;">
%s
</div>
<p style="margin:24px 0 0 0; font-size:14px; color:#666; text-align:center;">
This code expires in <strong>10 minutes</strong>. Please do not share it with anyone.
</p>
</td>
</tr>
<tr>
<td style="padding:0 40px 32px 40px;">
<p style="margin:0; font-size:13px; color:#888; text-align:center; line-height:1.5;">
If you did not create an account, you can safely ignore this email.
</p>
</td>
</tr>
<tr>
<td style="padding:24px 40px; background:#fafafe; border-top:1px solid #eee;">
<p style="margin:0; font-size:12px; color:#999; text-align:center;">
&copy; SmartCampus. All rights reserved.
</p>
</td>
</tr>
</table>
</td>
</tr>
</table>
</body>
</html>""" % (full_name, digit_spans)


async def send_otp_email(to_email: str, full_name: str, otp_code: str) -> None:
    """
    Asynchronously send the email-verification OTP email via SMTP.

    Fire-and-forget: every exception is caught and logged so that a delivery
    failure never propagates to the caller (mirrors Java's
    `catch (MessagingException) -> log` over @Async). The user can retry via
    the resend-OTP endpoint.

    No-op when MAIL_USERNAME is unset — enables dev/CI without SMTP creds.
    """
    if not settings.MAIL_USERNAME:
        logger.warning(
            "MAIL_USERNAME not set — skipping OTP email send to %s", to_email
        )
        return

    html = _build_otp_email_html(full_name, otp_code)

    message = EmailMessage()
    message["From"] = settings.MAIL_USERNAME
    message["To"] = to_email
    message["Subject"] = "Your SmartCampus Verification Code: " + otp_code
    message.set_content(html, subtype="html")

    try:
        await aiosmtplib.send(
            message,
            hostname="smtp.gmail.com",
            port=587,
            start_tls=True,
            username=settings.MAIL_USERNAME,
            password=settings.MAIL_PASSWORD,
        )
        logger.info("OTP email sent to %s", to_email)
    except Exception as e:  # noqa: BLE001 — intentional broad catch, see docstring
        logger.error("Failed to send OTP email to %s: %s", to_email, e)
        # Do NOT re-raise — fire-and-forget semantics (matches Java @Async + catch).
